tests/token/erc721/test_ERC721_SafeMintable_mock.py

```diff
@@ -0,0 +1,195 @@
+    state, account1, account2, erc721, erc721_holder, unsupported = erc721_init
+    _state = state.copy()
+    account1 = cached_contract(_state, account_def, account1)
+    account2 = cached_contract(_state, account_def, account2)
+    erc721 = cached_contract(_state, erc721_def, erc721)
+    erc721_holder = cached_contract(_state, erc721_holder_def, erc721_holder)
+    unsupported = cached_contract(_state, unsupported_def, unsupported)
+
+    return erc721, account1, account2, erc721_holder, unsupported
+
+
+@pytest.mark.asyncio
+async def test_safeMint_to_erc721_supported_contract(erc721_factory):
+    erc721, account, _, erc721_holder, _ = erc721_factory
+
+    await signer.send_transaction(
+        account, erc721.contract_address, 'safeMint', [
+            erc721_holder.contract_address,
+            *TOKEN,
+            len(DATA),
+            *DATA
+        ]
+    )
+
+    # check balance
+    execution_info = await erc721.balanceOf(erc721_holder.contract_address).call()
+    assert execution_info.result == (to_uint(1),)
+
+    # check owner
+    execution_info = await erc721.ownerOf(TOKEN).call()
+    assert execution_info.result == (erc721_holder.contract_address,)
+
+
+@pytest.mark.asyncio
+async def test_safeMint_to_account(erc721_factory):
+    erc721, account, recipient, _, _ = erc721_factory
+
+    await signer.send_transaction(
+        account, erc721.contract_address, 'safeMint', [
+            recipient.contract_address,
+            *TOKEN,
+            len(DATA),
+            *DATA
+        ]
+    )
+
+    # check balance
+    execution_info = await erc721.balanceOf(recipient.contract_address).call()
+    assert execution_info.result == (to_uint(1),)
+
+    # check owner
+    execution_info = await erc721.ownerOf(TOKEN).call()
+    assert execution_info.result == (recipient.contract_address,)
+
+
+@pytest.mark.asyncio
+async def test_safeMint_to_zero_address(erc721_factory):
+    erc721, account, _, _, _ = erc721_factory
+
+    # to zero address should be rejected
+    await assert_revert(signer.send_transaction(
+        account, erc721.contract_address, 'safeMint', [
+            ZERO_ADDRESS,
+            *TOKEN,
+            len(DATA),
+            *DATA
+        ]
+    ))
+
+
+@pytest.mark.asyncio
+async def test_safeMint_from_zero_address(erc721_factory):
+    erc721, _, _, erc721_holder, _ = erc721_factory
+
+    # Caller address is `0` when not using an account contract
+    await assert_revert(
+        erc721.safeMint(
+            erc721_holder.contract_address,
+            TOKEN,
+            DATA
+        ).invoke()
+    )
+
+
+@pytest.mark.asyncio
+async def test_safeMint_from_not_owner(erc721_factory):
+    erc721, _, other, erc721_holder, _ = erc721_factory
+
+    await assert_revert(signer.send_transaction(
+        other, erc721.contract_address, 'safeMint', [
+            erc721_holder.contract_address,
+            *TOKEN,
+            len(DATA),
+            *DATA
+        ])
+    )
```

Comment:
should we add revert messages in here or should we leave it for #190?

Reply:
I think we should leave it for #190 because there's an `Ownable_only_owner` assertion